src/hypothesis/stateful.py

```diff
@@ -354,13 +353,59 @@ def precondition_wrapper(*args, **kwargs):
             setattr(precondition_wrapper, PRECONDITION_MARKER, precond)
         else:
             new_rule = Rule(targets=rule.targets, arguments=rule.arguments,
-                            function=rule.function, precondition=precond,
-                            parent_rule=rule.parent_rule)
+                            function=rule.function, precondition=precond)
             setattr(precondition_wrapper, RULE_MARKER, new_rule)
+
+        invariant = getattr(f, INVARIANT_MARKER, None)
+        if invariant is not None:
+            new_invariant = Invariant(function=invariant.function,
+                                      precondition=precond)
+            setattr(precondition_wrapper, INVARIANT_MARKER, new_invariant)
+
         return precondition_wrapper
     return decorator
 
 
+Invariant = namedtuple(
+    u'Invariant',
+    (u'function', u'precondition')
+)
+
+
+def invariant():
```

Comment:
I'm not sure I agree - rule and precondition need to be called because they generally or always have arguments that need passing to them - but I'm happy to grant you author's privilege and leave it as you have it. :-)